README.md
# Simple golang file server

![Release version][badge_release_version]
![Project language][badge_language]
[![Build Status][badge_build]][link_build]
[![Coverage][badge_coverage]][link_coverage]
[![Go Report][badge_goreport]][link_goreport]
[![License][badge_license]][link_license]

This package provides basic file server functionality with:

- In memory caching with TTL and limits (like maximal cached files count and maximal file size)
- Overridable error handlers
- "Index" file serving (like `index` [nginx directive](http://nginx.org/en/docs/http/ngx_http_index_module.html#index))
- Redirection to the "parent" directory, when index file requested
- "Allowed methods" list

Most use-case is [SPA](https://en.wikipedia.org/wiki/Single-page_application) assets serving.

## Installation

To install package, use `go get`:

```bash
$ go get github.com/avto-dev/go-simple-fileserver
```

## Example

```go
fs, err := fileserver.NewFileServer(fileserver.Settings{
    FilesRoot:               "./web",
    IndexFileName:           "index.html",
    ErrorFileName:           "__error__.html",
    RedirectIndexFileToRoot: true,
    AllowedHTTPMethods:      []string{http.MethodGet},
    CacheEnabled:            true,
    CacheTTL:                time.Second * 5,
    CacheMaxFileSize:        1024 * 64, // 64 KiB
    CacheMaxItems:           512,
})

if err != nil {
    log.Fatal(err)
}

log.Fatal(http.ListenAndServe(":9000", fs))
```

To run this example execute `go run .` in `./examples` directory.

More information can be found in the godocs: <http://godoc.org/github.com/avto-dev/go-simple-fileserver>

### Testing

For package testing we use built-in golang testing feature and `docker-ce` + `docker-compose` as develop environment. So, just write into your terminal after repository cloning:

```shell
$ make test
```

## Changelog

[![Release date][badge_release_date]][link_releases]
[![Commits since latest release][badge_commits_since_release]][link_commits]

Changes log can be [found here][link_changes_log].

## Support

[![Issues][badge_issues]][link_issues]
[![Issues][badge_pulls]][link_pulls]

If you will find any package errors, please, [make an issue][link_create_issue] in current repository.

## License

This is open-sourced software licensed under the [MIT License][link_license].

[badge_build]:https://img.shields.io/github/workflow/status/avto-dev/go-simple-fileserver/tests/master
[badge_coverage]:https://img.shields.io/codecov/c/github/avto-dev/go-simple-fileserver/master.svg?maxAge=30
[badge_goreport]:https://goreportcard.com/badge/github.com/avto-dev/go-simple-fileserver
[badge_release_version]:https://img.shields.io/github/release/avto-dev/go-simple-fileserver.svg?maxAge=30
[badge_language]:https://img.shields.io/github/go-mod/go-version/avto-dev/go-simple-fileserver?longCache=true
[badge_license]:https://img.shields.io/github/license/avto-dev/go-simple-fileserver.svg?longCache=true
[badge_release_date]:https://img.shields.io/github/release-date/avto-dev/go-simple-fileserver.svg?maxAge=180
[badge_commits_since_release]:https://img.shields.io/github/commits-since/avto-dev/go-simple-fileserver/latest.svg?maxAge=45
[badge_issues]:https://img.shields.io/github/issues/avto-dev/go-simple-fileserver.svg?maxAge=45
[badge_pulls]:https://img.shields.io/github/issues-pr/avto-dev/go-simple-fileserver.svg?maxAge=45
[link_goreport]:https://goreportcard.com/report/github.com/avto-dev/go-simple-fileserver

[link_coverage]:https://codecov.io/gh/avto-dev/go-simple-fileserver
[link_build]:https://github.com/avto-dev/go-simple-fileserver/actions
[link_license]:https://github.com/avto-dev/go-simple-fileserver/blob/master/LICENSE
[link_releases]:https://github.com/avto-dev/go-simple-fileserver/releases
[link_commits]:https://github.com/avto-dev/go-simple-fileserver/commits
[link_changes_log]:https://github.com/avto-dev/go-simple-fileserver/blob/master/CHANGELOG.md
[link_issues]:https://github.com/avto-dev/go-simple-fileserver/issues
[link_create_issue]:https://github.com/avto-dev/go-simple-fileserver/issues/new/choose
[link_pulls]:https://github.com/avto-dev/go-simple-fileserver/pulls


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].

## v1.0.0

### Added

- In memory caching with TTL and limits (like maximal cached files count and maximal file size)
- Overridable error handlers
- "Index" file serving (like `index` [nginx directive](http://nginx.org/en/docs/http/ngx_http_index_module.html#index))
- Redirection to the "parent" directory, when index file requested
- "Allowed methods" list

[keepachangelog]:https://keepachangelog.com/en/1.0.0/
[semver]:https://semver.org/spec/v2.0.0.html


errors_test.go
package fileserver

import (
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestErrorPageTemplate_String(t *testing.T) {
	assert.Equal(t, "foo", ErrorPageTemplate("foo").String())
}

func TestErrorPageTemplate_Build(t *testing.T) {
	assert.Equal(t,
		"foo 200 <> OK",
		ErrorPageTemplate("foo {{ code }} <> {{ message }}").Build(200),
	)
}

func TestJSONErrorHandler(t *testing.T) {
	tmpDir, _ := ioutil.TempDir("", "test-")
	defer func(d string) { assert.NoError(t, os.RemoveAll(d)) }(tmpDir)

	fs, _ := NewFileServer(Settings{FilesRoot: tmpDir})
	assert.NotNil(t, fs)
	handler := JSONErrorHandler()

	var (
		req, _ = http.NewRequest(http.MethodGet, "", nil)
		rr     = httptest.NewRecorder()
	)

	assert.False(t, handler(rr, req, fs, http.StatusNotFound))

	req, _ = http.NewRequest(http.MethodGet, "", nil)
	req.Header.Add("Accept", "application/json")
	rr = httptest.NewRecorder()

	assert.True(t, handler(rr, req, fs, http.StatusNotFound))
	assert.Equal(t, "application/json; charset=utf-8", rr.Header().Get("Content-Type"))
	assert.JSONEq(t, `{"code":404,"message":"Not Found"}`, rr.Body.String())
}

func TestStaticHtmlPageErrorHandler(t *testing.T) {
	tmpDir, _ := ioutil.TempDir("", "test-")
	defer func(d string) { assert.NoError(t, os.RemoveAll(d)) }(tmpDir)

	cacheTTL := time.Millisecond * 5

	fs, _ := NewFileServer(Settings{
		FilesRoot:    tmpDir,
		CacheEnabled: true,
		CacheTTL:     cacheTTL,
	})
	assert.NotNil(t, fs)
	handler := StaticHTMLPageErrorHandler()

	var (
		req, _ = http.NewRequest(http.MethodGet, "", nil)
		rr     = httptest.NewRecorder()
	)

	assert.False(t, handler(rr, req, fs, http.StatusNotFound))

	// create template file
	file, _ := os.Create(filepath.Join(tmpDir, "error.html"))
	_, _ = file.Write([]byte("template: {{ message }} | {{ code }}"))
	file.Close()

	fs.Settings.ErrorFileName = "error.html"

	req, _ = http.NewRequest(http.MethodGet, "", nil)
	rr = httptest.NewRecorder()

	assert.True(t, handler(rr, req, fs, http.StatusBadGateway))
	assert.Equal(t, "text/html; charset=utf-8", rr.Header().Get("Content-Type"))
	assert.Equal(t, `template: Bad Gateway | 502`, rr.Body.String())

	// remove template file (cache must be used)
	assert.NoError(t, os.Remove(filepath.Join(tmpDir, "error.html")))

	rr = httptest.NewRecorder()

	assert.True(t, handler(rr, req, fs, http.StatusNotFound))
	assert.Equal(t, `template: Not Found | 404`, rr.Body.String())

	time.Sleep(cacheTTL)

	assert.False(t, handler(rr, req, fs, http.StatusNotFound))
}


fileserver_test.go
package fileserver

import (
	"io/ioutil"
	"math/rand"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func init() { //nolint:gochecknoinits
	rand.Seed(time.Now().UnixNano())
}

var letterRunes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ") //nolint:gochecknoglobals

func RandStringRunes(t *testing.T, n int) string {
	t.Helper()

	b := make([]rune, n)
	for i := range b {
		b[i] = letterRunes[rand.Intn(len(letterRunes))] //nolint:gosec
	}

	return string(b)
}

func TestNewFileServer_WrongDirectoryError(t *testing.T) {
	fs, err := NewFileServer(Settings{
		FilesRoot: RandStringRunes(t, 32),
	})

	assert.Nil(t, fs)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "not exists")

	tmpDir, _ := ioutil.TempDir("", "test-")
	defer func(d string) { assert.NoError(t, os.RemoveAll(d)) }(tmpDir)
	file, _ := os.Create(filepath.Join(tmpDir, "foo"))
	file.Close()

	fs, err = NewFileServer(Settings{
		FilesRoot: file.Name(),
	})

	assert.Nil(t, fs)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "not directory")
}

func TestFileServer_ServeHTTP(t *testing.T) {
	var cases = []struct {
		name                   string
		giveDirs               []string
		giveFiles              map[string][]byte
		giveSettings           Settings
		giveRequestMethod      string
		giveRequestURI         string
		giveRequestHeaders     map[string]string
		beforeServing          func(fs *FileServer)
		wantResponseHTTPCode   int
		wantResponseContent    string
		wantResponseSubstrings []string
		resultCheckingFn       func(t *testing.T, rr *httptest.ResponseRecorder)
	}{
		{
			name:                   "serving request without URI",
			giveRequestURI:         "",
			wantResponseHTTPCode:   http.StatusNotFound,
			wantResponseSubstrings: []string{"Not Found"},
		},
		{
			name:           "static file serving",
			giveRequestURI: "/test",
			giveFiles: map[string][]byte{
				"test": []byte("test content"),
			},
			wantResponseHTTPCode: http.StatusOK,
			wantResponseContent:  "test content",
		},
		{
			name: "disallowed HTTP method is used",
			giveSettings: Settings{
				AllowedHTTPMethods: []string{http.MethodPost},
			},
			giveRequestMethod:      http.MethodGet,
			giveRequestURI:         "/",
			wantResponseHTTPCode:   http.StatusMethodNotAllowed,
			wantResponseSubstrings: []string{"Method Not Allowed"},
		},
		{
			name: "disallowed HTTP method for existing file is used",
			giveSettings: Settings{
				AllowedHTTPMethods: []string{http.MethodPost, http.MethodDelete},
			},
			giveRequestMethod: http.MethodGet,
			giveRequestURI:    "/test",
			giveFiles: map[string][]byte{
				"test": []byte("test content"),
			},
			wantResponseHTTPCode:   http.StatusMethodNotAllowed,
			wantResponseSubstrings: []string{"Method Not Allowed"},
		},
		{
			name:                 "directory above (./../) requested",
			giveRequestURI:       "/../../../../etc/passwd",
			wantResponseHTTPCode: http.StatusNotFound,
		},
		{
			name: "disabled redirection from",
			giveSettings: Settings{
				IndexFileName: "idx.html",
			},
			giveRequestURI:       "/foo/idx.html",
			wantResponseHTTPCode: http.StatusNotFound,
		},
		{
			name: "redirect from ./{indexFileName} to ./",
			giveSettings: Settings{
				IndexFileName:           "idx.html",
				RedirectIndexFileToRoot: true,
			},
			giveRequestURI:       "/idx.html",
			wantResponseHTTPCode: http.StatusMovedPermanently,
			resultCheckingFn: func(t *testing.T, rr *httptest.ResponseRecorder) {
				assert.Equal(t, "/", rr.Header().Get("Location"))
			},
		},
		{
			name: "redirect from ./foo/{indexFileName} to ./foo/",
			giveSettings: Settings{
				IndexFileName:           "idx.html",
				RedirectIndexFileToRoot: true,
			},
			giveRequestURI:       "/foo/idx.html",
			wantResponseHTTPCode: http.StatusMovedPermanently,
			resultCheckingFn: func(t *testing.T, rr *httptest.ResponseRecorder) {
				assert.Equal(t, "/foo/", rr.Header().Get("Location"))
			},
		},
		{
			name: "index file in root directory serving",
			giveSettings: Settings{
				IndexFileName: "idx.html",
			},
			giveRequestURI: "/",
			giveFiles: map[string][]byte{
				"idx.html": []byte("index content"),
			},
			wantResponseHTTPCode: http.StatusOK,
			wantResponseContent:  "index content",
		},
		{
			name: "index file in sub-directory serving",
			giveSettings: Settings{
				IndexFileName: "idx.html",
			},
			giveRequestURI: "/foo/",
			giveDirs:       []string{"foo"},
			giveFiles: map[string][]byte{
				"idx.html":                       []byte("index in root"),
				filepath.Join("foo", "idx.html"): []byte("index in foo"),
			},
			wantResponseHTTPCode: http.StatusOK,
			wantResponseContent:  "index in foo",
		},
		{
			name: "404 on directory request",
			giveSettings: Settings{
				IndexFileName: "indx.html",
			},
			giveDirs: []string{"foo"},
			giveFiles: map[string][]byte{
				"indx.html":                       []byte("index in root"),
				filepath.Join("foo", "indx.html"): []byte("index in foo"),
			},
			giveRequestURI:       "/foo",
			wantResponseHTTPCode: http.StatusNotFound,
		},
		{
			name: "custom error handler",
			beforeServing: func(fs *FileServer) {
				fs.ErrorHandlers = []ErrorHandlerFunc{
					func(w http.ResponseWriter, r *http.Request, fs *FileServer, errorCode int) bool {
						w.WriteHeader(444)
						_, _ = w.Write([]byte("foo bar"))
						w.Header().Set("Content-Type", "blah blah")

						return true
					},
				}
			},
			giveRequestURI:       "/foo",
			wantResponseHTTPCode: 444,
			wantResponseContent:  "foo bar",
			resultCheckingFn: func(t *testing.T, rr *httptest.ResponseRecorder) {
				assert.Equal(t, "blah blah", rr.Header().Get("Content-Type"))
			},
		},
		{
			name: "custom error handler fallback",
			beforeServing: func(fs *FileServer) {
				fs.ErrorHandlers = []ErrorHandlerFunc{
					func(w http.ResponseWriter, r *http.Request, fs *FileServer, errorCode int) bool {
						return false
					},
				}
			},
			giveRequestURI:         "/foo",
			wantResponseHTTPCode:   http.StatusNotFound,
			wantResponseSubstrings: []string{"<html>", "Error 404", "Not Found", "</html>"},
		},
		{
			name:                 "error in json format when json requested",
			giveRequestURI:       "/foo",
			giveRequestHeaders:   map[string]string{"accept": "application/json"},
			wantResponseHTTPCode: http.StatusNotFound,
			resultCheckingFn: func(t *testing.T, rr *httptest.ResponseRecorder) {
				assert.JSONEq(t, `{"code":404,"message":"Not Found"}`, rr.Body.String())
			},
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			tmpDir, tmpDirErr := ioutil.TempDir("", "test-")
			assert.NoError(t, tmpDirErr)

			defer func(d string) { assert.NoError(t, os.RemoveAll(d)) }(tmpDir)

			if len(tt.giveDirs) > 0 || len(tt.giveFiles) > 0 {
				for _, d := range tt.giveDirs {
					assert.NoError(t, os.Mkdir(filepath.Join(tmpDir, d), 0777))
				}

				for name, content := range tt.giveFiles {
					file, createErr := os.Create(filepath.Join(tmpDir, name))
					assert.NoError(t, createErr)
					_, fileWritingErr := file.Write(content)
					assert.NoError(t, fileWritingErr)
					assert.NoError(t, file.Close())
				}
			}

			if tt.giveSettings.FilesRoot == "" {
				tt.giveSettings.FilesRoot = tmpDir
			}

			fs, fsErr := NewFileServer(tt.giveSettings)

			assert.NoError(t, fsErr)

			if tt.giveRequestMethod == "" { // setup default HTTP request method
				tt.giveRequestMethod = fs.Settings.AllowedHTTPMethods[0]
			}

			var (
				req, _ = http.NewRequest(tt.giveRequestMethod, tt.giveRequestURI, nil)
				rr     = httptest.NewRecorder()
			)

			if tt.giveRequestHeaders != nil {
				for key, value := range tt.giveRequestHeaders {
					req.Header.Set(key, value)
				}
			}

			if tt.beforeServing != nil {
				tt.beforeServing(fs)
			}

			fs.ServeHTTP(rr, req)

			//fmt.Println(rr.Code, rr.Body.String())

			assert.Equal(t, tt.wantResponseHTTPCode, rr.Code)

			if tt.wantResponseContent != "" {
				assert.Equal(t, tt.wantResponseContent, rr.Body.String())
			}

			if len(tt.wantResponseSubstrings) > 0 {
				for _, expected := range tt.wantResponseSubstrings {
					assert.Contains(t, rr.Body.String(), expected)
				}
			}

			if tt.resultCheckingFn != nil {
				tt.resultCheckingFn(t, rr)
			}
		})
	}
}

func TestFileServer_CacheUsage(t *testing.T) {
	// create dir
	tmpDir, _ := ioutil.TempDir("", "test-cache-")
	defer func(d string) { assert.NoError(t, os.RemoveAll(d)) }(tmpDir)

	// create files
	file, _ := os.Create(filepath.Join(tmpDir, "test"))
	_, _ = file.Write([]byte("test content"))
	file.Close()

	cacheTTL := time.Millisecond * 5

	fs, _ := NewFileServer(Settings{
		FilesRoot:     tmpDir,
		ErrorFileName: "error.html",
		CacheEnabled:  true,
		CacheTTL:      cacheTTL,
	})

	var (
		req, _ = http.NewRequest(http.MethodGet, "/test", nil)
		rr     = httptest.NewRecorder()
	)

	fs.ServeHTTP(rr, req)

	assert.Equal(t, http.StatusOK, rr.Code)
	assert.Equal(t, "test content", rr.Body.String()) // content from `./test`
	assert.NoError(t, os.Remove(filepath.Join(tmpDir, "test")))

	rr = httptest.NewRecorder()
	fs.ServeHTTP(rr, req)
	assert.Equal(t, http.StatusOK, rr.Code)
	assert.Equal(t, "test content", rr.Body.String()) // content from cache

	time.Sleep(cacheTTL) // cache must be expired after this line

	rr = httptest.NewRecorder()
	fs.ServeHTTP(rr, req)
	assert.Equal(t, http.StatusNotFound, rr.Code)
	assert.Contains(t, rr.Body.String(), "Not Found") // cache expired and now file not foud
}


cache/cache_test.go
package cache

import (
	"bytes"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestInMemoryCache_CRD(t *testing.T) {
	cache := NewInMemoryCache(time.Millisecond * 3)

	item, exists := cache.Get("foo")
	assert.Nil(t, item)
	assert.False(t, exists)
	assert.Equal(t, uint32(0), cache.Count())

	now := time.Now()
	data := []byte("abc")
	ttl := time.Millisecond * 7

	cache.Set("foo", ttl, &Item{
		ModifiedTime: now,
		Content:      bytes.NewReader(data),
	})

	item, exists = cache.Get("foo")
	assert.Equal(t, uint32(1), cache.Count())
	assert.Equal(t, now, item.ModifiedTime)

	buf := make([]byte, len(data))
	_, _ = item.Content.Read(buf)

	assert.Equal(t, []byte("abc"), buf)
	assert.True(t, exists)

	time.Sleep(ttl)

	item, exists = cache.Get("foo")
	assert.Nil(t, item)
	assert.False(t, exists)

	time.Sleep(ttl)

	assert.Equal(t, uint32(0), cache.Count())
}


errors.go
package fileserver

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"path"
	"strconv"
	"strings"
	"time"

	"github.com/avto-dev/go-simple-fileserver/cache"
)

// ErrorPageTemplate  is error page template in string representation. Is allowed to use basic "replacing patterns"
// like `{{ code }}` or `{{ message }}`
type ErrorPageTemplate string

// String converts template into string representation.
func (t ErrorPageTemplate) String() string { return string(t) }

// Build makes registered patterns replacing.
func (t ErrorPageTemplate) Build(errorCode int) string {
	out := t.String()

	for k, v := range map[string]string{
		"code":    strconv.Itoa(errorCode),
		"message": http.StatusText(errorCode),
	} {
		out = strings.ReplaceAll(out, fmt.Sprintf("{{ %s }}", k), v)
	}

	return out
}

type jsonError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// JSONErrorHandler respond with simple json-formatted response, if json format was requested (defined in `Accept`
// header).
func JSONErrorHandler() ErrorHandlerFunc {
	return func(w http.ResponseWriter, r *http.Request, fs *FileServer, errorCode int) bool {
		if strings.Contains(r.Header.Get("Accept"), "json") {
			w.Header().Set("Content-Type", "application/json; charset=utf-8")
			w.WriteHeader(errorCode)

			_ = json.NewEncoder(w).Encode(jsonError{
				Code:    errorCode,
				Message: http.StatusText(errorCode),
			})

			return true
		}

		return false
	}
}

// StaticHTMLPageErrorHandler allows to use user-defined local file with HTML for error page generating.
func StaticHTMLPageErrorHandler() ErrorHandlerFunc { //nolint:gocognit
	return func(w http.ResponseWriter, r *http.Request, fs *FileServer, errorCode int) bool {
		if len(fs.Settings.ErrorFileName) > 0 {
			var (
				filePath        = path.Join(fs.Settings.FilesRoot, fs.Settings.ErrorFileName)
				templateContent []byte
				loaded          bool
			)

			if fs.CacheAvailable() {
				if cached, cacheHit := fs.Cache.Get(filePath); cacheHit {
					templateContent, _ = ioutil.ReadAll(cached.Content)
					loaded = true
				}
			}

			if !loaded {
				if f, err := os.Open(filePath); err == nil {
					defer f.Close()

					if data, err := ioutil.ReadAll(f); err == nil {
						templateContent = data
						loaded = true

						if fs.CacheAvailable() && fs.Cache.Count() < fs.Settings.CacheMaxItems {
							fs.Cache.Set(filePath, fs.Settings.CacheTTL, &cache.Item{
								ModifiedTime: time.Now(),
								Content:      bytes.NewReader(data),
							})
						}
					}
				}
			}

			if loaded {
				w.Header().Set("Content-Type", "text/html; charset=utf-8")
				w.WriteHeader(errorCode)
				_, _ = w.Write([]byte(ErrorPageTemplate(templateContent).Build(errorCode)))

				return true
			}
		}

		return false
	}
}


cache/cache.go
package cache

import (
	"io"
	"time"

	"github.com/patrickmn/go-cache"
)

type (
	// Cacher interface describes something, who can read and write data into fast storage (faster than local
	// filesystem).
	Cacher interface {
		// Get an item from the cache. Returns the item or nil, and a bool indicating whether the key was found.
		Get(key string) (*Item, bool)

		// Set an item to the cache, replacing any existing item.
		Set(key string, ttl time.Duration, item *Item)

		// Count returns the number of items in the cache.
		Count() uint32
	}

	// Item is structured cache item.
	Item struct {
		ModifiedTime time.Time
		Content      io.ReadSeeker
	}
)

// InMemoryCache implements Cacher interface and uses memory as a storage.
type InMemoryCache struct {
	engine *cache.Cache
}

// NewInMemoryCache creates cacher implementation, that uses memory as a storage.
func NewInMemoryCache(cleanupInterval time.Duration) *InMemoryCache {
	return &InMemoryCache{
		engine: cache.New(cache.NoExpiration, cleanupInterval),
	}
}

// Get an item from the cache. Returns the item or nil, and a bool indicating whether the key was found.
func (c *InMemoryCache) Get(key string) (*Item, bool) {
	item, ok := c.engine.Get(key)

	if item == nil {
		return nil, ok
	}

	return item.(*Item), ok
}

// Set an item to the cache, replacing any existing item. If the duration is -1, the item never expires.
func (c *InMemoryCache) Set(key string, ttl time.Duration, item *Item) {
	c.engine.Set(key, item, ttl)
}

// Count returns the number of items in the cache. This may include items that have expired, but have not yet been
// cleaned up.
func (c *InMemoryCache) Count() uint32 {
	return uint32(c.engine.ItemCount())
}


fileserver.go
package fileserver

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"path"
	"path/filepath"
	"strings"
	"time"

	"github.com/avto-dev/go-simple-fileserver/cache"
)

const (
	defaultFallbackErrorContent = "<html><body><h1>Error {{ code }}</h1><h2>{{ message }}</h2></body></html>"
	defaultIndexFileName        = "index.html"
	defaultCacheTTL             = time.Second * 5
	defaultCacheMaxFileSize     = 1024 * 64 // 64 KiB
	defaultCacheMaxItems        = 64
)

// ErrorHandlerFunc is used as handler for errors processing. If func return `true` - next handler will be NOT executed.
type ErrorHandlerFunc func(w http.ResponseWriter, r *http.Request, fs *FileServer, errorCode int) (doNotContinue bool)

// FileServer is a main file server structure (implements `http.Handler` interface).
type FileServer struct {
	// Server settings (some of them can be changed in runtime).
	Settings Settings

	// Cacher instance.
	Cache cache.Cacher // nil, if caching disabled

	// If all error handlers fails - this content will be used as fallback for error page generating.
	FallbackErrorContent string

	// Error handlers stack.
	ErrorHandlers []ErrorHandlerFunc

	// Allowed HTTP methods map (is used in performance reasons).
	allowedHTTPMethodsMap map[string]struct{} // fillable in runtime
}

// Settings describes file server options.
type Settings struct {
	// Directory path, where files for serving is located.
	FilesRoot string

	// File name (relative path to the file) that will be used as an index (like <https://bit.ly/356QeFm>).
	IndexFileName string

	// File name (relative path to the file) that will be used as error page template.
	ErrorFileName string

	// Respond "index file" request with redirection to the root (`example.com/index.html` -> `example.com/`).
	RedirectIndexFileToRoot bool

	// Allowed HTTP methods (eg.: `http.MethodGet`).
	AllowedHTTPMethods []string

	// Enables caching engine.
	CacheEnabled bool

	// Maximal data caching lifetime.
	CacheTTL time.Duration

	// Maximum file size (in bytes), that can be placed into the cache.
	CacheMaxFileSize int64

	// Maximum files count, that can be placed into the cache.
	CacheMaxItems uint32
}

// NewFileServer creates new file server with default settings. Feel free to change default behavior.
func NewFileServer(s Settings) (*FileServer, error) {
	if info, err := os.Stat(s.FilesRoot); err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf(`directory "%s" does not exists`, s.FilesRoot)
		}

		return nil, err
	} else if !info.IsDir() {
		return nil, fmt.Errorf(`"%s" is not directory`, s.FilesRoot)
	}

	if s.IndexFileName == "" {
		s.IndexFileName = defaultIndexFileName
	}

	if s.CacheTTL == 0 {
		s.CacheTTL = defaultCacheTTL
	}

	if s.CacheMaxFileSize == 0 {
		s.CacheMaxFileSize = defaultCacheMaxFileSize
	}

	if s.CacheMaxItems == 0 {
		s.CacheMaxItems = defaultCacheMaxItems
	}

	if len(s.AllowedHTTPMethods) == 0 {
		s.AllowedHTTPMethods = append(s.AllowedHTTPMethods, http.MethodGet)
	}

	fs := &FileServer{
		Settings:             s,
		FallbackErrorContent: defaultFallbackErrorContent,
	}

	if s.CacheEnabled {
		fs.Cache = cache.NewInMemoryCache(s.CacheTTL / 2) //nolint:gomnd
	}

	fs.ErrorHandlers = []ErrorHandlerFunc{
		JSONErrorHandler(),
		StaticHTMLPageErrorHandler(),
	}

	return fs, nil
}

// CacheAvailable checks cache availability.
func (fs *FileServer) CacheAvailable() bool {
	return fs.Settings.CacheEnabled && fs.Cache != nil
}

func (fs *FileServer) handleError(w http.ResponseWriter, r *http.Request, errorCode int) {
	if fs.ErrorHandlers != nil && len(fs.ErrorHandlers) > 0 {
		for _, handler := range fs.ErrorHandlers {
			if handler(w, r, fs, errorCode) {
				return
			}
		}
	}

	// fallback
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.WriteHeader(errorCode)

	_, _ = w.Write([]byte(ErrorPageTemplate(fs.FallbackErrorContent).Build(errorCode)))
}

func (fs *FileServer) methodIsAllowed(method string) bool {
	if fs.allowedHTTPMethodsMap == nil {
		// burn allowed methods map for fast checking
		fs.allowedHTTPMethodsMap = make(map[string]struct{})

		for _, v := range fs.Settings.AllowedHTTPMethods {
			fs.allowedHTTPMethodsMap[v] = struct{}{}
		}
	}

	_, found := fs.allowedHTTPMethodsMap[method]

	return found
}

// ServeHTTP responds to an HTTP request.
func (fs *FileServer) ServeHTTP(w http.ResponseWriter, r *http.Request) { //nolint:funlen,gocognit,gocyclo
	if !fs.methodIsAllowed(r.Method) {
		fs.handleError(w, r, http.StatusMethodNotAllowed)

		return
	}

	if fs.Settings.RedirectIndexFileToRoot && len(fs.Settings.IndexFileName) > 0 {
		// redirect .../index.html to .../
		if strings.HasSuffix(r.URL.Path, "/"+fs.Settings.IndexFileName) {
			http.Redirect(w, r, r.URL.Path[0:len(r.URL.Path)-len(fs.Settings.IndexFileName)], http.StatusMovedPermanently)

			return
		}
	}

	urlPath := r.URL.Path

	// add leading `/` (if required)
	if len(urlPath) == 0 || !strings.HasPrefix(urlPath, "/") {
		urlPath = "/" + r.URL.Path
	}

	// if directory requested (or server root) - add index file name
	if len(fs.Settings.IndexFileName) > 0 && urlPath[len(urlPath)-1] == '/' {
		urlPath += fs.Settings.IndexFileName
	}

	// prepare target file path
	filePath := path.Join(fs.Settings.FilesRoot, filepath.FromSlash(path.Clean(urlPath)))

	// look for response in cache
	if fs.CacheAvailable() {
		if cached, cacheHit := fs.Cache.Get(filePath); cacheHit {
			http.ServeContent(w, r, filepath.Base(filePath), cached.ModifiedTime, cached.Content)

			return
		}
	}

	// check for file existence
	if stat, err := os.Stat(filePath); err == nil && stat.Mode().IsRegular() {
		if file, err := os.Open(filePath); err == nil {
			defer file.Close()

			var fileContent io.ReadSeeker

			// put file content into cache, if it is possible
			if fs.CacheAvailable() &&
				fs.Cache.Count() < fs.Settings.CacheMaxItems &&
				stat.Size() <= fs.Settings.CacheMaxFileSize {
				if data, err := ioutil.ReadAll(file); err == nil {
					fileContent = bytes.NewReader(data)

					fs.Cache.Set(filePath, fs.Settings.CacheTTL, &cache.Item{
						ModifiedTime: stat.ModTime(),
						Content:      fileContent,
					})
				}
			}

			if fileContent == nil {
				fileContent = file
			}

			http.ServeContent(w, r, filepath.Base(filePath), stat.ModTime(), fileContent)

			return
		}

		fs.handleError(w, r, http.StatusInternalServerError)

		return
	}

	fs.handleError(w, r, http.StatusNotFound)
}
